controllers/druid/drain_historical.go

```diff
@@ -0,0 +1,191 @@
+
+		if obj.(*appsv1.StatefulSet).Status.ReadyReplicas != obj.(*appsv1.StatefulSet).Status.CurrentReplicas {
+			msg := fmt.Sprintf("StatefulSet[%s] roll out is in progress CurrentReplicas[%d] != ReadyReplicas[%d]", nodeSpecUniqueStr, obj.(*appsv1.StatefulSet).Status.CurrentReplicas, obj.(*appsv1.StatefulSet).Status.ReadyReplicas)
+			logger_drain_historical.Info(msg, "name", m.Name, "namespace", m.Namespace)
+			return nil
+		}
+
+		m.Status.Historical.Replica = obj.(*appsv1.StatefulSet).Status.CurrentReplicas
+		m.Status.Historical.CurrentBatch = 0
+	}
+
+	originalReplicaCount := m.Status.Historical.Replica - batchSize
+
+	podList, err := readers.List(context.TODO(), sdk, m, makeLabelsForNodeSpec(nodeSpec, m, m.Name, nodeSpecUniqueStr), emitEvent, func() objectList { return &v1.PodList{} }, func(listObj runtime.Object) []object {
+		items := listObj.(*v1.PodList).Items
+		result := make([]object, len(items))
+		for i := 0; i < len(items); i++ {
+			result[i] = &items[i]
+		}
+		return result
+	})
+	if err != nil {
+		return err
+	}
+
+	pvcLabels := map[string]string{
+		"druid_cr":          m.Name,
+		"nodeSpecUniqueStr": nodeSpecUniqueStr,
+		"component":         nodeSpec.NodeType,
+	}
+
+	pvcList, err := readers.List(context.TODO(), sdk, m, pvcLabels, emitEvent, func() objectList { return &v1.PersistentVolumeClaimList{} }, func(listObj runtime.Object) []object {
+		items := listObj.(*v1.PersistentVolumeClaimList).Items
+		result := make([]object, len(items))
+		for i := 0; i < len(items); i++ {
+			result[i] = &items[i]
+		}
+		return result
+	})
+	if err != nil {
+		return err
+	}
+
+	podNames := getPodNames(podList)
+	//start draining the historical pod in batches and delete them
+	batchCount := originalReplicaCount/nodeSpec.DeploymentConfig.BatchSize + 1
+
+	//TODO add auth
+	d := InitClient(baseUrl, userName, os.Getenv("DRUID_PASSWORD"))
+	CoordinatorConfig, err := d.GetCoordinatorConfig()
+
+	if err != nil {
+		return err
+	}
+	startPod := m.Status.Historical.CurrentBatch * nodeSpec.DeploymentConfig.BatchSize
+	endPod := startPod + nodeSpec.DeploymentConfig.BatchSize - 1
+	if m.Status.Historical.CurrentBatch == batchCount {
+		endPod = originalReplicaCount - 1
+	}
+	if m.Status.Historical.DecommissionedPods == nil {
+		for podId := startPod; podId <= endPod; podId++ {
+			m.Status.Historical.DecommissionedPods = append(m.Status.Historical.DecommissionedPods, podNames[podId])
+		}
+		CoordinatorConfig["decommissioningNodes"] = []string(m.Status.Historical.DecommissionedPods) //TODO fix this
+		err := d.UpdateCoordinatorConfig(CoordinatorConfig)
+		if err != nil {
+			return err
+		}
+	}
+	//wait for draining to complete so we can start with the next batch
+	drainStatus, err := checkDrainStatus(d, startPod, endPod, podNames)
+	if err != nil {
+		return err
+	}
+	if !drainStatus {
+		logger_drain_historical.Info("Waiting for pods to drain", "name", m.Name, "namespace", m.Namespace)
+		return nil
+	}
+	//delete corresponding nodegrabbers
```

Comment:
This can be implemented only if we use local storage. Which suggests that we need some kind of check to see if its a local storage instance or not.